src/rt-app.c

```diff
@@ -413,6 +432,70 @@ static int run_event(event_data_t *event, int dry_run,
 			pthread_yield();
 		}
 		break;
+	case rtapp_fork:
+		{
+			log_debug("fork %d", event->count);
+
+			/*
+			 * If multiple threads race to fork, we must ensure
+			 * each one sees a unique index. Hence the lock.
+			 */
+			pthread_mutex_lock(&fork_mutex);
+			int new_thread_index = nthreads++;
+			threads = reallocarray(threads, nthreads, sizeof(*threads));
```

Comment:
Do you think we can pre-calculate the total size of the array (including instance 0 tasks) from the beginning?
It should be more efficient at runtime no?

Reply:
Static allocation was my first approach actually. I had a hard limit on 1024 forks only to minimize allocation (hence uncertainty in execution time). But when I had a chat with Chris about this at some point in the past he had a valid point that when we fork there's an overhead anyway so it could be considered as intrinsic part of the behaviour of this type of loads. Loads that fork have an associated uncertainty of what happens next when the execution of the parent resumes.

I could change it back to static but reallocarray() will not be _that_ slow (don't quote me on that) in the context of the systems rt-app is usually ran on.

Reply:
Keep in mind since we can have nested loops (the forked tasks has a phase that forks which in return forks a task that hsa phase that forks etc) - it might get ugly to try to calculate that.

Reply:
OK. Not a big deal for the moment. Feel free to keep it as is.